File: common-core/src/main/java/io/quartic/geojson/GeometryVisitor.java
package io.quartic.geojson;

public interface GeometryVisitor<T> {
    T visit(Polygon polygon);
    T visit(Point point);
    T visit(LineString lineString);

    T visit(MultiLineString multiLineString);
    T visit(MultiPoint multiPoint);
    T visit(MultiPolygon multiPolygon);

    default T visit(Geometry geometry) {
        throw new UnsupportedOperationException("Cannot process type " + geometry.getClass().getCanonicalName());
    }
}

File: common-core/src/main/kotlin/io/quartic/geojson/GeometryVisitor.kt
package io.quartic.geojson

interface GeometryVisitor<T> {
    fun visit(polygon: Polygon): T
    fun visit(point: Point): T
    fun visit(lineString: LineString): T

    fun visit(multiLineString: MultiLineString): T
    fun visit(multiPoint: MultiPoint): T
    fun visit(multiPolygon: MultiPolygon): T

    fun visit(geometry: Geometry): T {
        throw UnsupportedOperationException("Cannot process type " + geometry.javaClass.canonicalName)
    }
}
